### PR TITLE
[tcling] Suppress `-Wunused-result` diagnostics in wrappers generated by TClingCallFunc

### LINKED ISSUE #8622: PyROOT triggers a warning about [[nodiscard]] vector::empty with GCC11
### Describe the bug
```python
import ROOT

ROOT.gInterpreter.Declare(
    """
    struct S{
        std::vector<int> vec1;
    };
    """)

s = ROOT.S()
if (s.vec1.empty()):
    print("s vec1 is empty")
```
```
$ python struct_with_vec_empty_nodiscard.py 
input_line_38:10:7: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
      ((const vector<int>*)obj)->empty();
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
s vec1 is empty
```
### Expected behavior

No warning triggered.

### Setup
ROOT 6.25 built on Fedora34 with GCC 11

### Additional context
This is the behaviour at the root prompt
```
root [0] std::vector<int> vec1;
root [1] vec1.empty()
ROOT_prompt_1:1:1: warning: ignoring return value of function declared with 'nodiscard' attribute [-Wunused-result]
vec1.empty()
^~~~~~~~~~
(bool) true
root [2] if (vec1.empty()) {}
```

### PR DESCRIPTION
**Pull request by jalopezg-r00t**
This pull-request suppresses `-Wunused-result` diagnostics for wrapper functions generated by TClingCallFunc (see below).

A TClingCallFunc wrapper function might look as the excerpt below, where the function denoted by `func` may have been annotated as `[[nodiscard]]`. Note that if `ret == nullptr` the result of the call is unused.
```c++
extern "C" void __cf_0(void* obj, int nargs, void** args, void* ret) {
   if (ret) {
      new (ret) (return_type) ((class_name*)obj)->func(args...);
   }
   else {
      ((class_name*)obj)->func(args...);
   }
}
```

In turn, this triggers warnings when used by cppyy/PyROOT, e.g.
```python
>>> import ROOT
>>> v = ROOT.std.vector(int)()
>>> v.empty()
input_line_34:10:7: warning; ignoring return value of function declared with 'nodiscard' attribute
      [-Wunused-result]
      ((const vector<int>*)obj)->empty();
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
True
>>>
```

## Changes or fixes:
- Given the above situation, this commit supresses `-Wunused-result` diagnostics only for TClingCallFunc wrapper functions.

## Checklist:
- [X] tested changes locally

This PR fixes #8622.

Sibling PR in roottest: https://github.com/root-project/roottest/pull/791

### CODE REVIEW

**Review by vepadulano — APPROVED** [review]
LGTM, thanks for finding this out it !

**Comment by eguiraud** [comment]
Would changing 

```
((class_name*)obj)->func(args...);
```

to 

```
(void)((class_name*)obj)->func(args...);
```

also fix it? (I can't judge if that's a better or worse solution :) )

**Comment by pcanal** [comment]
I think Enrico's solution might be slight better as it is a more local solution (and clarify the intent).

**Comment by jalopezg-r00t** [comment]
> Would changing
> 
> ```
> ((class_name*)obj)->func(args...);
> ```
> 
> to
> 
> ```
> (void)((class_name*)obj)->func(args...);
> ```
> 
> also fix it? (I can't judge if that's a better or worse solution :) )

Yes, of course.  Changed!  Thanks for the suggestion, @eguiraud!

**Review by vgvassilev — COMMENTED** [review]
Can we get a test for this, presumably in https://github.com/root-project/root/blob/master/core/metacling/test/TClingCallFuncTests.cxx

File: core/metacling/test/TClingCallFuncTests.cxx [hunk 1]
```diff
@@ -14,6 +16,19 @@
 // system, feel free to delete them as these tests here don't represent things the user
 // should do in his code.
 
+class FilterDiagsRAII {
```
Comment (vgvassilev) [inline core/metacling/test/TClingCallFuncTests.cxx:19]: Thanks. That seems quite useful. Can you add it in this header https://github.com/root-project/root/blob/master/test/unit_testing_support/ROOTUnitTestSupport.h

**Review by vgvassilev — APPROVED** [review]
Lgtm!

**Pull request merged**